streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -177,6 +177,14 @@ void handleRebalanceComplete() {
         // before then the assignment has not been updated yet.
         mainConsumer.pause(mainConsumer.assignment());
 
+        if (stateUpdater != null) {
+            // All tasks that need restoration are now owned by the state updater.
+            // All tasks that are owned by the task manager are ready and can be resumed immediately.
+            for (final Task t : tasks.allTasks()) {
+                mainConsumer.resume(t.inputPartitions());
+            }
+        }
```

Comment:
Could we do something like:

```
final Set<TopicPartition> partitionsNotToPause = tasks.allTasks().stream().flatMap(task -> task.inputPartitions().stream()).collect(Collectors.toSet());
mainConsumer.pause(mainConsumer.assignment().stream().filter(partition -> !partitionsNotToPause.contains(partition)).collect(Collectors.toSet()));
```
or

```
final Set<TopicPartition> partitionsNotToPause = tasks.allTasks().stream().flatMap(task -> task.inputPartitions().stream()).collect(Collectors.toSet());
final Set<TopicPartition> partitionsToPause = new HashSet<>(mainConsumer.assignment());
partitionsToPause.removeAll(partitionsNotToPause);
mainConsumer.pause(partitionsToPause);        
```

Reply:
+1. The consumer.pause/resume calls are not completely free so it's better to batch them.

Reply:
Personally I prefer the second, i.e. just pause what's necessary; in the past we simply pause all since all of them need to go through a restoration phase.